samza-core/src/main/java/org/apache/samza/zk/ZkBarrier.java

```diff
@@ -0,0 +1,214 @@
+   * @param participants List of expected participated for this barrier to complete
+   */
+  public void create(final String version, List<String> participants) {
+    String barrierRoot = keyBuilder.getBarrierRoot();
+    zkUtils.makeSurePersistentPathsExists(new String[]{
+        barrierRoot,
+        keyBuilder.getBarrierPath(version),
+        keyBuilder.getBarrierParticipantsPath(version),
+        keyBuilder.getBarrierStatePath(version)});
+
+    // subscribe for participant's list changes
+    String barrierParticipantsPath = keyBuilder.getBarrierParticipantsPath(version);
+    LOG.info("Subscribing for child changes at " + barrierParticipantsPath);
+    zkUtils.getZkClient().subscribeChildChanges(barrierParticipantsPath, new ZkBarrierChangeHandler(version, participants));
+
+    if (barrierListener != null) {
+      barrierListener.onBarrierCreated(version);
+    }
+  }
+
+  /**
+   * Joins a shared barrier by registering under the barrier sub-tree in ZK
+   *
+   * @param version Version associated with the Barrier
+   * @param participantId Identifier of the participant
+   */
+  public void join(String version, String participantId) {
+    String barrierDonePath = keyBuilder.getBarrierStatePath(version);
+    zkUtils.getZkClient().subscribeDataChanges(barrierDonePath, new ZkBarrierReachedHandler(barrierDonePath, version));
+
+    // TODO: Handle ZkNodeExistsException
+    zkUtils.getZkClient().createPersistent(
+        String.format("%s/%s", keyBuilder.getBarrierParticipantsPath(version), participantId));
+  }
+
+  /**
+   * Expires the barrier version by marking it as TIMED_OUT
+   *
+   * @param version Version associated with the Barrier
+   */
+  public void expire(String version) {
+    zkUtils.getZkClient().writeData(
+        keyBuilder.getBarrierStatePath(version),
+        State.TIMED_OUT);
+
+  }
+  /**
+   * Listener for changes to the list of participants. It is meant to be subscribed only by the creator of the barrier
+   * node. It checks to see when the barrier is ready to be marked as completed.
+   */
+  class ZkBarrierChangeHandler implements IZkChildListener {
+    private final String barrierVersion;
+    private final List<String> names;
+
+    public ZkBarrierChangeHandler(String barrierVersion, List<String> names) {
+      this.barrierVersion = barrierVersion;
+      this.names = names;
+    }
+
+    @Override
+    public void handleChildChange(String parentPath, List<String> currentChildren) {
+      if (currentChildren == null) {
+        LOG.info("Got ZkBarrierChangeHandler handleChildChange with null currentChildren");
+        return;
+      }
+      LOG.debug("list of children in the barrier = " + parentPath + ":" + Arrays.toString(currentChildren.toArray()));
+      LOG.debug("list of children to compare against = " + parentPath + ":" + Arrays.toString(names.toArray()));
+
+      // check if all the expected participants are in
+      if (CollectionUtils.containsAll(currentChildren, names)) {
```

Comment:
Correct me if I am wrong, `currentChildren` at the end of barrier will have all the participants right? 
If yes,  a quick check on the size as an additional condition might be performant since `Collectionutils.containsAll(...)` doesn't check for sizes.